File: app-kotlin/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/activities/investigation/evidence/EvidenceFragment.java
package com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.evidence;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.annotation.SuppressLint;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewStub;
import android.widget.FrameLayout;
import android.widget.PopupWindow;
import android.widget.RelativeLayout;
import android.widget.ScrollView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.constraintlayout.widget.ConstraintLayout;

import com.TritiumGaming.phasmophobiaevidencepicker.R;
import com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.InvestigationFragment;
import com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.evidence.views.DifficultyCarouselView;
import com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.evidence.views.MapCarouselView;
import com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.evidence.views.PhaseTimerControlView;
import com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.evidence.views.PhaseTimerView;
import com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.evidence.views.SanitySeekBarView;
import com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.evidence.views.SanityMeterView;
import com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.evidence.views.evidence.EvidenceList;
import com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.evidence.views.ghost.GhostList;
import com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.evidence.views.investigation.InvestigationSection;
import com.TritiumGaming.phasmophobiaevidencepicker.listeners.CompositeListener;
import com.TritiumGaming.phasmophobiaevidencepicker.views.investigation.sanity.SanityWarningView;

/**
 * EvidenceFragment class
 *
 * @author TritiumGamingStudios
 */
public class EvidenceFragment extends InvestigationFragment {

    protected InvestigationSection ghostSection, evidenceSection;

    protected GhostList ghostList;
    protected EvidenceList evidenceList;

    protected ConstraintLayout sanityTrackingConstraintLayout;

    protected AppCompatImageView toggleCollapseButton;
    protected AppCompatTextView phaseTimerTextView;
    protected AppCompatTextView sanityPercentTextView;

    protected CompositeListener compositeListenerPrev, compositeListenerNext;

    protected DifficultyCarouselView difficultyCarouselView;
    protected PhaseTimerView phaseTimerCountdownView;
    protected PhaseTimerControlView playPauseButton;
    protected MapCarouselView mapTrackControl;
    protected SanitySeekBarView sanitySeekBarView;
    protected SanityMeterView sanityMeterView;
    protected SanityWarningView sanityWarningTextView;


    /**
     * EvidenceFragment constructor
     * <p>
     * TODO
     *
     * @param layout -
     */
    public EvidenceFragment(int layout) {
        super(layout);
    }

    @Nullable
    @Override
    public View onCreateView(
            @NonNull LayoutInflater inflater,
            @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState) {

        return inflater.inflate(R.layout.fragment_evidence_solo, container, false);
    }

    @SuppressLint("ClickableViewAccessibility")
    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {

        super.onViewCreated(view, savedInstanceState);

        if(evidenceViewModel.getSanityData() != null) {
            evidenceViewModel.getSanityData()
                    .setFlashTimeoutMax(globalPreferencesViewModel.getHuntWarningFlashTimeout());
        }

        // GHOST / EVIDENCE CONTAINERS
        FrameLayout
                column_left = view.findViewById(R.id.column_left),
                column_right = view.findViewById(R.id.column_right);
        column_right.findViewById(R.id.scrollview)
                .setVerticalScrollbarPosition(View.SCROLLBAR_POSITION_RIGHT);

        if(!globalPreferencesViewModel.isLeftHandSupportEnabled()) {
            ghostSection = (InvestigationSection) column_left.getChildAt(0);
            evidenceSection = (InvestigationSection) column_right.getChildAt(0);
        } else {
            evidenceSection = (InvestigationSection) column_left.getChildAt(0);
            ghostSection = (InvestigationSection) column_right.getChildAt(0);
        }

        ghostSection.setLabel(getString(R.string.investigation_section_title_ghosts));
        evidenceSection.setLabel(getString(R.string.investigation_section_title_evidence));

        ScrollView ghost_scrollview = ghostSection.findViewById(R.id.scrollview);
        ScrollView evidence_scrollview = evidenceSection.findViewById(R.id.scrollview);

        ghostList = new GhostList(requireContext());
        evidenceList = new EvidenceList(requireContext());

        ghostList.init(
                globalPreferencesViewModel, evidenceViewModel,
                popupWindow, ghostSection.findViewById(R.id.progressbar),
                adRequest);
        evidenceList.init(
                globalPreferencesViewModel, evidenceViewModel,
                popupWindow, evidenceSection.findViewById(R.id.progressbar),
                adRequest, ghostList);

        ViewStub list_ghosts = ghostSection.findViewById(R.id.list);
        ViewStub list_evidence = evidenceSection.findViewById(R.id.list);

        ghost_scrollview.removeView(list_ghosts);
        ghost_scrollview.addView(ghostList);

        evidence_scrollview.removeView(list_evidence);
        evidence_scrollview.addView(evidenceList);

        //ghostContainer.requestDisallowInterceptTouchEvent(true);

        toggleCollapseButton = view.findViewById(R.id.button_toggleSanity);

        // TIMER VIEW
        phaseTimerTextView = view.findViewById(R.id.evidence_timer_text);

        // SANITY METER VIEWS
        sanityPercentTextView = view.findViewById(R.id.evidence_sanitymeter_percentage);
        sanityMeterView = view.findViewById(R.id.evidence_sanitymeter_progressbar);
        sanitySeekBarView = view.findViewById(R.id.evidence_sanitymeter_seekbar);
        sanityWarningTextView = view.findViewById(R.id.evidence_sanitymeter_huntwarning);

        // SANITY COLLAPSIBLE
        sanityTrackingConstraintLayout = view.findViewById(R.id.constraintLayout_sanityTracking);

        if(toggleCollapseButton != null) {
            toggleCollapseButton.setOnClickListener(v -> {
                if(evidenceViewModel.isDrawerCollapsed) {
                    sanityTrackingConstraintLayout.animate()
                            .setListener(new AnimatorListenerAdapter() {
                                @Override
                                public void onAnimationStart(Animator animation) {
                                    super.onAnimationStart(animation);
                                    sanityTrackingConstraintLayout.setVisibility(View.GONE);
                                }

                                @Override
                                public void onAnimationEnd(Animator animation) {
                                    super.onAnimationEnd(animation);

                                    sanityTrackingConstraintLayout.setVisibility(View.VISIBLE);
                                    toggleCollapseButton.setImageLevel(2);
                                    evidenceViewModel.isDrawerCollapsed = false;
                                }
                            })
                            .start();
                } else {
                    sanityTrackingConstraintLayout.animate()
                            .setListener(new AnimatorListenerAdapter() {
                                @Override
                                public void onAnimationStart(Animator animation) {
                                    super.onAnimationStart(animation);

                                    sanityTrackingConstraintLayout.setVisibility(View.VISIBLE);
                                }
                                @Override
                                public void onAnimationEnd(Animator animation) {
                                    super.onAnimationStart(animation);

                                    sanityTrackingConstraintLayout.setVisibility(View.GONE);
                                    toggleCollapseButton.setImageLevel(1);
                                    evidenceViewModel.isDrawerCollapsed = true;
                                }
                            })
                            .start();
                }
            });

            initCollapsible();
        }

        // SANITY
        if (sanitySeekBarView != null) {
            sanitySeekBarView.init(
                    evidenceViewModel.getSanityData(),
                    sanityPercentTextView);
            sanitySeekBarView.resetProgress();
        }

        sanityMeterView.init(evidenceViewModel.getSanityData());

        popupWindow = new PopupWindow(
                RelativeLayout.LayoutParams.MATCH_PARENT,
                RelativeLayout.LayoutParams.MATCH_PARENT);

        ghostList.createPopupWindow(popupWindow);
        evidenceList.createPopupWindow(popupWindow);
        new Thread(() -> ghostList.createViews()).start();
        new Thread(() -> evidenceList.createViews()).start();

    }

    private void initCollapsible() {
        if(evidenceViewModel.isDrawerCollapsed) {
            sanityTrackingConstraintLayout.setVisibility(View.GONE);
            toggleCollapseButton.setImageLevel(1);
        } else {
            sanityTrackingConstraintLayout.setVisibility(View.VISIBLE);
            toggleCollapseButton.setImageLevel(2);
        }
    }

    @Override
    public void softReset() {
        if (evidenceViewModel != null) {
            evidenceViewModel.reset();
        }

        if (sanitySeekBarView != null) {
            sanitySeekBarView.updateProgress();
        }

        if(phaseTimerCountdownView != null) {
            phaseTimerCountdownView.destroyTimer();
        }
    }

    public void requestInvalidateComponents() {

        if(evidenceViewModel != null) {
            evidenceViewModel.getGhostOrderData().updateOrder();
        }

        if(evidenceList != null) {
            //evidenceList.forceResetEvidenceContainer();
        }

        if(ghostList != null) {
            ghostList.forceResetGhostContainer();
        }

        // SANITY
        if (sanitySeekBarView != null) {
            sanitySeekBarView.resetProgress();
        }

        if(evidenceViewModel != null &&
                evidenceViewModel.getPhaseTimerData() != null) {
            playPauseButton.pause();
            phaseTimerCountdownView.reset();
        }

        if(sanityWarningTextView != null) {
            sanityWarningTextView.reset();
        }

    }

    /**
     * onPause
     */
    @Override
    public void onPause() {

        phaseTimerCountdownView.destroyTimer();

        super.onPause();
    }

    /**
     * onDestroyView
     */
    @Override
    public void onDestroyView() {

        if (sanityMeterView != null) {
            sanityMeterView.recycleBitmaps();
        }

        if(popupWindow != null) {
            popupWindow.dismiss();
            popupWindow = null;
        }

        super.onDestroyView();
    }

    /**
     * onResume
     */
    @Override
    public void onResume() {

        if (!sanityMeterView.hasBuiltImages()) {
            sanityMeterView.buildImages();
        }

        super.onResume();
    }

    @Override
    protected void saveStates() {

    }

}

File: app-java/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/activities/investigation/evidence/EvidenceFragment.kt
package com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.evidence

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.annotation.SuppressLint
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewStub
import android.widget.FrameLayout
import android.widget.PopupWindow
import android.widget.RelativeLayout
import android.widget.ScrollView
import androidx.appcompat.widget.AppCompatImageView
import androidx.constraintlayout.widget.ConstraintLayout
import com.TritiumGaming.phasmophobiaevidencepicker.R
import com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.InvestigationFragment
import com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.evidence.views.investigation.sanity.SanityToolsLayout
import com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.evidence.views.investigation.section.InvestigationSection
import com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.evidence.views.investigation.section.lists.EvidenceListView
import com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.evidence.views.investigation.section.lists.GhostListView

open class EvidenceFragment(layout: Int) : InvestigationFragment(layout) {
    protected var ghostSection: InvestigationSection? = null
    protected var evidenceSection: InvestigationSection? = null

    protected var ghostList: GhostListView? = null
    protected var evidenceList: EvidenceListView? = null

    protected var toggleCollapseButton: AppCompatImageView? = null
    protected var sanityTrackingConstraintLayout: ConstraintLayout? = null

    protected var sanityToolsLayout: SanityToolsLayout? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_evidence, container, false)
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        sanityToolsLayout = view.findViewById(R.id.layout_sanity_tool)
        sanityToolsLayout?.init(evidenceViewModel)

        evidenceViewModel.sanityModel?.flashTimeoutMax = 
            globalPreferencesViewModel.huntWarningFlashTimeout.toLong()

        // GHOST / EVIDENCE CONTAINERS
        val columnLeft = view.findViewById<FrameLayout>(R.id.column_left)
        val columnRight = view.findViewById<FrameLayout>(R.id.column_right)
        columnRight.findViewById<View>(R.id.scrollview).verticalScrollbarPosition =
            View.SCROLLBAR_POSITION_RIGHT

        if (!globalPreferencesViewModel.isLeftHandSupportEnabled) {
            ghostSection = columnLeft.getChildAt(0) as InvestigationSection
            evidenceSection = columnRight.getChildAt(0) as InvestigationSection
        } else {
            evidenceSection = columnLeft.getChildAt(0) as InvestigationSection
            ghostSection = columnRight.getChildAt(0) as InvestigationSection
        }

        ghostSection?.setLabel(getString(R.string.investigation_section_title_ghosts))
        evidenceSection?.setLabel(getString(R.string.investigation_section_title_evidence))

        val ghostScrollview = ghostSection?.findViewById<ScrollView>(R.id.scrollview)
        val evidenceScrollview = evidenceSection?.findViewById<ScrollView>(R.id.scrollview)

        ghostList = GhostListView(requireContext())
        evidenceList = EvidenceListView(requireContext())

        ghostList?.init(
            globalPreferencesViewModel, evidenceViewModel,
            popupWindow, ghostSection?.findViewById(R.id.progressbar),
            adRequest
        )
        evidenceList?.init(
            globalPreferencesViewModel, evidenceViewModel,
            popupWindow, evidenceSection?.findViewById(R.id.progressbar),
            adRequest, ghostList
        )

        val listGhosts = ghostSection?.findViewById<ViewStub>(R.id.list)
        val listEvidence = evidenceSection?.findViewById<ViewStub>(R.id.list)

        ghostScrollview?.removeView(listGhosts)
        ghostScrollview?.addView(ghostList)

        evidenceScrollview?.removeView(listEvidence)
        evidenceScrollview?.addView(evidenceList)

        toggleCollapseButton = view.findViewById(R.id.button_toggleSanity)

        // SANITY COLLAPSIBLE
        sanityTrackingConstraintLayout = view.findViewById(R.id.constraintLayout_sanityTracking)

        if (toggleCollapseButton != null) {
            toggleCollapseButton?.setOnClickListener {
                if (evidenceViewModel.isDrawerCollapsed) {
                    sanityTrackingConstraintLayout?.animate()
                        ?.setListener(object : AnimatorListenerAdapter() {
                            override fun onAnimationStart(animation: Animator) {
                                super.onAnimationStart(animation)

                                sanityTrackingConstraintLayout?.visibility = View.GONE
                            }

                            override fun onAnimationEnd(animation: Animator) {
                                super.onAnimationEnd(animation)

                                sanityTrackingConstraintLayout?.visibility = View.VISIBLE
                                toggleCollapseButton?.setImageLevel(2)
                                evidenceViewModel.isDrawerCollapsed = false
                            }
                        })
                        ?.start()
                } else {
                    sanityTrackingConstraintLayout?.animate()
                        ?.setListener(object : AnimatorListenerAdapter() {
                            override fun onAnimationStart(animation: Animator) {
                                super.onAnimationStart(animation)

                                sanityTrackingConstraintLayout?.visibility = View.VISIBLE
                            }

                            override fun onAnimationEnd(animation: Animator) {
                                super.onAnimationStart(animation)

                                sanityTrackingConstraintLayout?.visibility = View.GONE
                                toggleCollapseButton?.setImageLevel(1)
                                evidenceViewModel.isDrawerCollapsed = true
                            }
                        })
                        ?.start()
                }
            }

            initCollapsible()
        }

        popupWindow = PopupWindow(
            RelativeLayout.LayoutParams.MATCH_PARENT,
            RelativeLayout.LayoutParams.MATCH_PARENT
        )

        ghostList?.createPopupWindow(popupWindow)
        evidenceList?.createPopupWindow(popupWindow)
        Thread { ghostList?.createViews() }.start()
        Thread { evidenceList?.createViews() }.start()
    }

    private fun initCollapsible() {
        if (evidenceViewModel.isDrawerCollapsed) {
            sanityTrackingConstraintLayout?.visibility = View.GONE
            toggleCollapseButton?.setImageLevel(1)
        } else {
            sanityTrackingConstraintLayout?.visibility = View.VISIBLE
            toggleCollapseButton?.setImageLevel(2)
        }
    }

    override fun reset() {
        if (evidenceViewModel != null) {
            evidenceViewModel.reset()
        }

        // TODO Force progress bar update

        // TODO Reset and Pause PhaseTimer
    }

    fun requestInvalidateComponents() {
        if (evidenceViewModel != null) {
            evidenceViewModel.ghostOrderData?.updateOrder()
        }

        if (ghostList != null) {
            ghostList?.forceResetGhostContainer()
        }


        // TODO Force progress bar update (aka reset)


        // TODO Reset Play/Pause button (to 'Play' state)\
    }

    override fun onDestroyView() {
        if (popupWindow != null) {
            popupWindow?.dismiss()
            popupWindow = null
        }

        super.onDestroyView()
    }

    override fun onResume() {
        super.onResume()
    }

    override fun saveStates() {
    }
}
